apps/blog-app/src/content/2026-07-12-logichat-qa-to-doc-rag-training.md

```
@@ -0,0 +1,297 @@
+draft: false
+---
+
+LogiChat is a chatbot platform. Customers upload their docs, get a chat widget, never touch a model. For two years, "training the bot" meant hand-curating a Q&A list in a dashboard form — `question` and `answer` pairs, one row at a time, dumped into the prompt as few-shot pairs. We just deleted that and replaced it with document-based retrieval. This post is about the architecture we landed on, the mistakes we made along the way, and the one piece of the old system we kept as a fallback.
+
+The full change touched five issues across four Cloud Run services. The interesting parts are all in two files: `apps/agent/src/agent/agent_definition.py` and `apps/agent/src/agent/retrieval.py`.
+
+## The old way: Q&A as few-shot prompt
+
+The original training flow was a Firestore collection per app:
+
+```
```

markdownlint-cli2 (0.22.1)

[warning] Fenced code blocks should have a language specified

(MD040, fenced-code-language)